src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(copyState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const keyR of action.keysToRemove) {
+        delete copyState[keyR];
+      }
+    }
+
+    if (action.type === 'clear') {
+      Object.keys(copyState).forEach((key) => delete copyState[key]);
+    }
+
+    states.push({ ...copyState });
```

Comment:
It is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This is already done correctly here, but ensure it remains outside of any conditional blocks such as `if` or `switch`.